### ISSUE TITLE
Network requests are being performed without the user's consent

### ISSUE THREAD

**Issue by hecktest** [minetest/minetest#14829]
### Minetest version

```true
5.8.0
```

### Operating system and version

all

### Summary

Split from #14819

Making the CDB update request from the main menu on startup is bad because it unintentionally becomes a tracking beacon that collects activity data. 

A second source of unwanted requests is automatic updates; most programs ask the user at some point whether they want to do this.

Network requests shouldn't be performed without user action or consent, there's multiple reasons to avoid this:

- Downstream repositories might not be happy about it
- User trust will be damaged and MT's reputation could suffer
- Server resources are wasted
- User resources (mobile data) are being wasted, if the phone is roaming on another continent it can be very costly
- This could be illegal in some jurisdictions, now or in the future (EU most likely)
- A vulnerability vector in the checker could affect more users before it's discovered

Solution: require reasonable user action before a network request is made. Any automatic action must be an opt-in setting.

I also suggest that consent collection is delegated to installers to minimize the nag factor. Certain distributions (windows ZIP) should not enable any collection or do any nagging by default. When building from source, the behavior should also be opt-in without nagging.

Recommended policy:
- Source distributions, default build settings: **Opt-in, no nag**
- Binary ZIP: **Opt-in, no nag**
- Binary installer: **Opt-in, ask in the installer**
- Mobile OS package: **Opt-in, formspec modal dialog on first run and the first time the CDB browser is used**

### Steps to reproduce

Launch the client. A network request will immediately be made.

**Comment by Desour** [comment]
Apart from ContentDB update checking and engine update checks, I'd like to add that there's also the request to the public server list as soon as you open the online tab. And the user can't avoid it (easily), even if they just want to join a server via an address that they got from somewhere else, or that is in their "favourites" server list.

----

>I also suggest that consent collection is delegated to installers to minimize the nag factor.

Imo it would be nicer UX to have something to decide the first time you use a feature, and not all at once at installation. (Plus having to test all the different installers could be an issue.)
Also, not opting in shouldn't make the update checker feature inaccessible.
For example, it would be nice to have a button to do each update checker feature manually, plus a checkbox next to it for a setting that enables automatic checking.

**Comment by sfan5** [comment]
> * Downstream repositories might not be happy about it

What we should do foremost is document what exactly MT does. So that downstreams can make an informed decision instead of maybe finding out about things when they stumble upon them.

> * User resources (mobile data) are being wasted, if the phone is roaming on another continent it can be very costly

I think this is sufficiently addressed by rate-limiting/caching (see #14433).
If you use MT for 12h every day for a whole month straight and trigger the checks at every opportunity you will end up at about 11 MB data usage. This is with 30% added for protocol overhead and assuming CDB does *not* do compression. 

<i>replies to stuff mentioned in the original issue</i>

> I don't know if a repo should ping for updates like this, no decent software repository I know of does this unprompted. Pacman, pkg_add, apt (probably) and others all require manual action to perform an update, and will download the new index opportunistically when they need it. 

This is a poor example because I know that Fedora (incl. derivatives like RHEL) and Ubuntu do fetch package repositories periodically by default. Arch added some automatic [GPG key sync](https://gitlab.archlinux.org/archlinux/archlinux-keyring/-/commit/ad8698e96c423dfc68405b547f310f2e1075a95d) a while ago. At least the GUI versions of openSUSE also poll for updates to let you know.
Sure, more security-consicous distros will not do this. But I think it's safe to say even in the FOSS world "software should not check for updates by default" is not consensus.

> Zip versions and source distributions should never phone home and probably not even nag the user about it. 

ZIP is still the default distribution way on Windows so that wouldn't make a lot of sense.
"Source distributions" also includes all the Linux distros as well.

> Do mods have version compat manifests? If not, fix that. Create safer ways for users to discover that they need to check for updates instead of hammering a central server.

Most updates are because of new features/fixes, not because of an incompatibility that could be detected via static checks.

> Right now, my personal safety could depend on your server's security, and I didn't even know about this until yesterday. 

I see two takeaways:
* We should sandbox the mainmenu lua env, just in case
* Users should be running Minetest in a sandbox more often. They should not have to trust us.

Sandboxing is unfortunately in a relatively poor state on desktop systems.

<hr>

I think automatically checking for MT updates has value for the average user. Because the average user is exactly the kind of user who will not click on a conveniently placed "check for updates" button. So I would enable it by default.
I'm neutral on whether CDB should check for updates on startup or only in the tab.

Either way there should of course be an option to disable any such checks, #14827 takes care of that.

Now for **consent**: I would solve this by having a dialog on first-launch before any requests are made.
Short description and a link to the privacy policy.
Accept -> enable CDB and MT update checks; Reject -> Don't and continue

If we're being pedantic we could also add a second dialog that's shown before opening the CDB or Server tab *if* the user didn't accept for the first time.

And speaking of dev builds it does happen that normal players use these and I think having an one-time privacy dialog is acceptable. So the behavior shouldn't differ from release builds.

**Comment by rubenwardy** [comment]
The ContentDB update detection feature was added because very few people would know to open the ContentDB dialog to see updates to their installed content. They'd just start a game and not even be aware that their content was very outdated. Update detection was one of the prerequisites to debundling Minetest Game

If a dialog is added, update detection needs to be the preferred choice for users that just click through dialogs without caring

Perhaps a privacy conscious alternative to update detection would be to show a "there might be updates" badge if the user hasn't opened the ContentDB dialog in a month, could have a button to fetch the updates list

**Comment by celeron55** [comment]
My opinion is that a one-time dialog should be shown where the user has to choose whether to enable or disable automatic updates.

It should be a neutral choice with two buttons to make sure we're not pushing users either way. After that, they will never be bothered about it again, unless they clear their minetest.conf. It should of course be toggleable in the settings afterwards.

The best time to show it would be right after the user has installed something from cdb, which is the thing that will trigger automatic updates if automatic updates are enabled. This will avoid it feeling naggy, as it is asked right when the question becomes relevant. Asking it at startup when it's unknown whether the user will ever even enter the content tab to begin with would be stupid.

I wouldn't like to see this be an opt-in nor an opt-out thing, because it's a valuable feature to many users, and because privacy is an equally valuable feature to just as many users.

**Comment by sfan5** [comment]
> You seem to have an extremely low opinion of your "average" users. Maybe they won't click it because they can live without it? I'd say the majority would not opt in when given the chance, but is also too lazy or unaware to opt out.

I think the average Minetest user is much closer to the average gamer than a FOSS enthusiast and that automatic update checks are right for them. Underlying reason being what @rubenwardy said.

Having a first-run dialog that asks you to opt into a recommend set of automatic stuff is not disrespecting power users.

> Would you go to a store where an employee performs a flash photography mugshot of you every time you enter and asks you to participate in a survey? If not, then why do you put up with it online?

I don't know what this analogy is supposed to allude to. But perhaps as something fitting if a website asked me to provide my email address to *read* something I'd leave in 10/10 cases.

**Comment by hlqkj** [comment]
I am not sure it would matter in this context, but requests made to media servers (I'm referring to the `remote_media` setting) are also made without the user knowing it.

Recently, I discovered by chance that a minetest server I regularly play on uses such a media server, after noting an error in the client logs which also revealed the URL, totally unrelated and unknown to me: can't say it wasn't a little disappointing...

True that we have the `enable_remote_media_server` setting, but there's a few facts that make it subtle:

- It is enabled by default and it is easily forgotten (like in my case).
- Even though the user has control on whether to enable or disable this feature, they are not informed on where their requests will be directed; most of the times the media server is not the server they are connecting to.
- There is no way to selectively enable or disable this feature for specific servers.

**Comment by rubenwardy** [comment]
> in doing so, it sends stats about what Windows/MT version I use to the ContentDB - a feature I don't use at all 

The request is only made if you've installed something using the ContentDB menu integration

> I still don't know if the same information that gets sent to the ContentDB has also been included in requests to the media server.

It's part of the User-Agent so the same data is sent

**Comment by hlqkj** [comment]
> The request is only made if you've installed something using the ContentDB menu integration. So if you don't use the ContentDB feature at all, no requests will be made

Thanks for the clarification. However, I still think no one would ever think their client *and* system data is sent for stats if they choose to use the ContentDB - nowhere that is stated (AFAIK?). So the point about user trust and MT reputation till stands.

**Comment by appgurueu** [comment]
> I am not sure it would matter in this context, but requests made to media servers (I'm referring to the `remote_media` setting) are also made without the user knowing it.

I think this matters much, especially after #14851.

When you connect to a server, you know that network requests are being made, and that ultimately that server decides where your data ends up. The server could just be a proxy forwarding your data to an untrustworthy server. If you don't trust the server (and that includes their choice of media server), don't connect.

**Comment by hlqkj** [comment]
We're not talking about the player's best practices, what you said it's obvious (sorry). Whatever a player responsibility is, it is best practice and IMHO due for Minetest to inform its users and ask their consent where appropriate.

Secondly, even after #14851: why is the engine sending *unnecessary* data at all? For a specific service, only strictly required data to provide such service shall be sent. Adding more for stat purposes calls for user information and consent even more strongly (what this issue is all about).

The motif "Don't want to send your data? Don't use the service" doesn't sound so FOSS-friendly to me.

**Comment by appgurueu** [comment]
> Whatever a player responsibility is, it is best practice and IMHO due for Minetest to inform its users and ask their consent where appropriate.

Yes, and I do not think it is appropriate to ask for consent *again* here. We would just be nagging users to solve an issue that doesn't exist or is mostly irrelevant.

Roughly, connecting to a server is analogous to visiting a website in your browser. When the user explicitly makes such an action, that *is* consent to sending the data needed for the website (server) to work (well). And these websites will use CDNs - roughly the equivalent of remote media servers - and your browser will not ask you for explicit consent to connect to each of those CDNs, and that is fine and expected.

> why is the engine sending unnecessary data at all? For a specific service, only strictly required data to provide such service shall be sent

The problem is "strictly required" isn't well defined. There are definitely legitimate use cases for including version numbers in user agents, statistics being one of them. (That said, I don't have strong feelings about excluding this information, since I feel it isn't too useful. But I'm also not hosting a server.)

> what this issue is all about

Disagree. This issue is about network requests being performed *without the user's consent*. Specifically, even if you were just playing and working on content locally, Minetest would make network requests with very specific information - such as your installed packages - and I understand why users have a problem with that.

> The motif "Don't want to send your data? Don't use the service" doesn't sound so FOSS-friendly to me.

Said motif sounds like a trivial tautology to me given any networked service.

**Comment by hlqkj** [comment]
> Yes, and I do not think it is appropriate to ask for consent _again_ here (with regards to connecting to servers / remote media servers). We would just be nagging users to solve an issue that doesn't exist or is mostly irrelevant.

Agreed. As in my first comment, I just meant to mention here that there are also requests to media servers, unknown to the user, and as rubenwardy confirmed above those requests contain the same User-Agent.

> Roughly, connecting to a server is analogous to visiting a website in your browser. When the user explicitly makes such an action, that is consent to sending the data needed for the website (server) to work (well). And these websites will use CDNs - roughly the equivalent of remote media servers - and your browser will not ask you for explicit consent to connect to each of those CDNs, and that is fine and expected.

By the same website analogy, that is all what the Same-Origin Policy is about, and most recent browsers (Minetest is the equivalent in this analogy) block wild cross-origin requests, and enforce policies to those they allow. In general, browsers do what I have been trying to express here the whole time: that is, granted that everyone is responsible for their actions, adopt policies and implementations that are aimed at protecting the average user. Not throwing features on the table and then get over issues like this and #14830 letting the whole thing on the kids playing Minetest.

> The problem is "strictly required" isn't well defined. 

It *is* well defined instead: one just has to answer simple questions, e.g.:

- Do I need to know what OS the user is running, its version, system architecture, etc. to let them download/update a mod?
- Do I need to know the same to simply send a file from a content server?

You mention stats. Those are not a strict requirement either, but may be a "legitimate interest" in that they allow a service provider to diagnose and improve their service. However, even legitimate interest use mandates that the user is offered an option to oppose their data bein used.

Lastly, technically speaking, the User-Agent is not even a mandatory header (and we could debate if it did more good or harm in the past days) which makes me wonder why does Minetest enforces a default one at all, and hardcoded.

> Said motif sounds like a trivial tautology to me given any networked service.

Call it a tautology then, and on the fact that any service is networked now, include our full sys info, including unique IDs on the Minetest default User-Agent.

**Comment by rubenwardy** [comment]
I don't think the problem here is contacting the remote media server. If the Minetest connection packet contained the platform info, you'd have the same level of information disclosure here

> most recent browsers (Minetest is the equivalent in this analogy) block wild cross-origin requests,

This is not true, you can load images from another domain by default without cross origin issues. Browsers limit cross-origin requests made from scripts - but the requests can be allowed by the cross origin server.

> include our full sys info, including unique IDs on the Minetest default User-Agent - without mentioning this anywhere but in the source code.

It's not the full sys info and it does not include unique IDs. Here are some examples:

* `Minetest/5.9.0 (Linux/6.9.7)`
* `Minetest/5.8.0 (Android/32 aarch64)`
* `Minetest/5.7.0 (Windows/10.0.19041 x86_64)`

**Comment by hlqkj** [comment]
> It's not the full sys info and it does not include unique IDs. Here are some examples:

Thanks @rubenwardy, I thought it was clear enough that my comment was pure sarcasm - I am aware those are not being sent.

Also, sorry in that some parts of my comment are off-topic here, being more related to #14830 or too focused on replying others' comments.

I'll stop because, frankly, I clearly can't seem to express what I wanted to say here, since everyone keeps hitting back to single words and commas.

I thought the concept was simple: do favor respecting user privacy and right to be informed, give them full control and knowledge of what their computer does. Failing to do this, and we could easily have been in a situation in which we had those UIDs in the User-Agent, before someone like @hecktest noticed them in the code and opened these issues!

**Comment by sfan5** [comment]
> I said they block **wild** requests (-> e.g. made from scripts) and enforce policies on others (-> can be allowed). This is _really_ looking at the finger.

I think you're still misunderstanding what browsers do about third-party requests.
On any browser - in the default configuration - I can import media resources from anywhere freely and the browser will make a request on the user's behalf to that server, with no extra consent dialogs.
[SOP](https://developer.mozilla.org/en-US/docs/Web/Security/Same-origin_policy) concerns itself mostly with what can be done *with* the returned data. What browsers are increasingly doing is blocking third-party *cookies*.

> I thought the concept was simple: do favor respecting user privacy and right to be informed, give them full control and knowledge of what their computer does.

Our [privacy policy](https://www.minetest.net/app-privacy-policy/) is exactly the right document for that.
I agree that we should ensure that users have a chance to read and agree to it before MT sends any requests.

**Comment by hlqkj** [comment]
> Our [privacy policy](https://www.minetest.net/app-privacy-policy/) is exactly the right document for that.
> I agree that we should ensure that users have a chance to read and agree to it before MT sends any requests.

I may be wrong on this, if so I apologize. But that's the problem indeed: I went through the whole download and "install" procedure again and I found no reference to that privacy policy, neither in the engine nor in the website - except for the link in the website footer, although it is very hard to find, and hence to know about its very existence, unless one is specifically looking for it.

In addition to that I have to again mention that in years using Minetest I never thought I'd have to bother about data collection(*) or hidden requests, mainly because I was confident that being FOSS software it would not reserve such surprises (trust - such an overrated value nowadays). My mistake.

Finally, the privacy policy covers for the engine itself, server list, and contentdb, as it should be. It can't and shouldn't, of course, account for the servers you connect to, which may or may not inform the user about what they actually do with their data. For this very specific reason I have been advocating here that the engine should have been more careful about what it allows servers and mods to do, and what it includes in requests (e.g. including an unnecessary default User-Agent). And by the way, this doesn't only apply to, as by doing HTTP requests in my server-side mods I was not aware my server info would be included in the request.

(*) Before someone points this out: here I am not talking about data which is _obviously_ collected being the essence of the game itself: map data, signs and books text, chat, whatever...

**Comment by red-001** [comment]
> Agreed. As in my first comment, I just meant to mention here that there are also requests to media servers, unknown to the user, and as rubenwardy confirmed above those requests contain the same User-Agent.

The media server is set by the game server, or do you object to passing the user-agent to the game server as well? 

It seems to me this can't be easily solved with a setting, need to either make the decision to fully remove the user-agent and version string that's sent to the server, or keep them and accept this as just something that's required of a good client.

Having a setting that can be set to send a blank version/user-agent string just makes the clients with that set more traceable if it defaults to being disabled - and if it defaults to being enabled - well that would make the statistics useless.
